File: q61_Rotate List/solution.py
'''
Given a linked list, rotate the list to the right by k places, where k is non-negative.

Example 1:

Input: 1->2->3->4->5->NULL, k = 2
Output: 4->5->1->2->3->NULL
Explanation:
rotate 1 steps to the right: 5->1->2->3->4->NULL
rotate 2 steps to the right: 4->5->1->2->3->NULL
'''

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def rotateRight(self, head: ListNode, k: int) -> ListNode:
        last = head
        if not head:
            return(None)
        length = 1
        while(last.next):
            last = last.next
            length+=1
        last.next = head
        k = k%length
        end = head
        for i in range(length - k - 1):
            end = end.next
        start = end.next
        end.next =None
        return(start)

File: q61_Rotate List/solution.java
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null){return head;}
        int length = 1;
        ListNode end = head;
        while(end.next != null){
            end = end.next;
            length = length +1;
        }
        k = k%length;
        end.next = head;
        for(int i=0;i<(length-1-k);i++){
            head = head.next;
        }
        ListNode cur = head.next;
        head.next = null;
        return cur;
        
    }
}
